internal/server/cookie.go

```diff
@@ -2,42 +2,66 @@ package server
-	CookieLoginName         = "login"
-	CookieDomain            = ""
-	CookiePath              = "/"
-	// while these vars look goofy, they avoid "magic number" arguments to SetCookie
-	CookieHTTPOnlyNotJavascriptAccessible = true    // setting HttpOnly to true means JS can't access it.
-	CookieSecureHTTPSOnly                 = true    // setting Secure to true means the cookie is only sent over https connections
-	CookieMaxAgeDeleteImmediately         = int(-1) // <0: delete immediately
-	CookieMaxAgeNoExpiry                  = int(0)  // zero has special meaning of "no expiry"
+const (
+	cookieAuthorizationName       = "auth"
+	cookieLoginName               = "login"
+	cookiePath                    = "/"
+	cookieMaxAgeDeleteImmediately = -1 // <0: delete immediately
+	cookieMaxAgeNoExpiry          = 0  // zero has special meaning of "no expiry"
 )
 
 func setAuthCookie(c *gin.Context, key string, expires time.Time) {
 	maxAge := int(time.Until(expires).Seconds())
-	if maxAge == CookieMaxAgeNoExpiry {
-		maxAge = CookieMaxAgeDeleteImmediately
+	if maxAge == cookieMaxAgeNoExpiry {
+		maxAge = cookieMaxAgeDeleteImmediately
 	}
 
-	secure := CookieSecureHTTPSOnly
+	secure := true
 	if c.Request.TLS == nil {
 		// if the request came over HTTP, then the cookie will need to be sent unsecured
 		secure = false
 	}
 
-	c.SetSameSite(http.SameSiteStrictMode)
-
-	c.SetCookie(CookieAuthorizationName, key, maxAge, CookiePath, CookieDomain, secure, CookieHTTPOnlyNotJavascriptAccessible)
-	c.SetCookie(CookieLoginName, "1", maxAge, CookiePath, CookieDomain, secure, CookieHTTPOnlyNotJavascriptAccessible)
+	http.SetCookie(c.Writer, &http.Cookie{
+		Name:     cookieAuthorizationName,
+		Value:    url.QueryEscape(key),
+		MaxAge:   maxAge,
+		Path:     cookiePath,
+		SameSite: http.SameSiteStrictMode,
+		Secure:   secure,
+		HttpOnly: true, // not accessible by javascript
+	})
+	http.SetCookie(c.Writer, &http.Cookie{
+		Name:     cookieLoginName,
+		Value:    "1",
+		MaxAge:   maxAge,
+		Path:     cookiePath,
+		SameSite: http.SameSiteStrictMode,
+		Secure:   secure,
+		HttpOnly: true, // not accessible by javascript
+	})
 }
 
-func deleteAuthCookie(c *gin.Context) {
-	c.SetCookie(CookieAuthorizationName, "", CookieMaxAgeDeleteImmediately, CookiePath, CookieDomain, CookieSecureHTTPSOnly, CookieHTTPOnlyNotJavascriptAccessible)
-	c.SetCookie(CookieLoginName, "", CookieMaxAgeDeleteImmediately, CookiePath, CookieDomain, CookieSecureHTTPSOnly, CookieHTTPOnlyNotJavascriptAccessible)
+func deleteAuthCookie(writer http.ResponseWriter) {
+	http.SetCookie(writer, &http.Cookie{
+		Name:     cookieAuthorizationName,
+		MaxAge:   cookieMaxAgeDeleteImmediately,
+		Path:     cookiePath,
+		Secure:   true, // only over https
+		HttpOnly: true, // not accessible by javascript
+	})
+
+	http.SetCookie(writer, &http.Cookie{
+		Name:     cookieLoginName,
+		MaxAge:   cookieMaxAgeDeleteImmediately,
+		Path:     cookiePath,
+		Secure:   true, // only over https
```

Comment:
It seems strange that we set `secure: true` for deleting a cookie, but we don't require it for setting a cookie. If we can set a cookie on non-https, don't we also need to delete it on non-https? Otherwise the cookie will still be there for non-https?

Should these changes to `secure: false`, since they are deleting a sensitive value (not saving one).

Reply:
it probably needs to match the behavior of the setting function. I think delete calls set cookie under the hood

Reply:
Not sure what you mean. Are you confirming that we should?

This is clearly calling `http.SetCookie`, no question there.  Above this in  `setAuthCookie` is where we create the cookies. When we create them we don't always set `Secure: true`. We based it on if the request came in over http or https.

So my understanding is that a login request that comes in over `http` will receive the session cookie, but a logout over `http` won't remove it from the client, which seems strange and probably not right.

I'd expect these calls to `http.SetCookie` in `deleteAuthCookie` to always use `Secure: false`, since there is nothing sensitive in the values to delete a cookie from the client.

Reply:
I'm saying the setting and deleting of the cookie need to use the same `secure: true` or `secure: false` setting for both. I suspect deleting a cookie with secure: true won't remove a cookie set with secure: false, and/or vice versa

Reply:
Reading https://datatracker.ietf.org/doc/html/rfc6265#section-5.2.5 it appears this `secure` setting enables the `secure-only-flag` in the cookie car, which does this ([ref](https://datatracker.ietf.org/doc/html/rfc6265#section-5.4)):

> If the cookie's secure-only-flag is true, then the request uri's scheme must denote a "secure" protocol (as defined by the user agent).

My understanding is this means that a cookie with `secure: true` would only be sent back to the server if the url starts with `https://`.

I didn't see anything about this flag being used to compare cookies in the cookie jar, so I'm not sure if it's considered, or if the cookie name is used exclusively.



--- 

I believe we allow http for users who want to front the service with their own LB or proxy that handles TLS for them. We should definitely warn if the `http` addr is not localhost.

Reply:
https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 seems to cover storage, and doesn't mention considering the secure flag. So I think setting `secure: false` for delete of cookies might be correct.